internal/driver/onvifdiscovery.go

```diff
@@ -209,3 +218,106 @@ func executeRawProbe(conn net.Conn, params netscan.Params) ([]onvif.Device, erro
+	deviceMap := make(map[string]contract.Device, len(devices))
+
+	for _, dev := range devices {
+		if dev.Name == d.serviceName {
+			// skip control plane device
+			continue
+		}
+
+		onvifInfo, ok := dev.Protocols[OnvifProtocol]
+		if !ok {
+			d.lc.Warnf("Found registered device %s without %s protocol information.", dev.Name, OnvifProtocol)
+			continue
+		}
+
+		endpointRef := onvifInfo[EndpointRefAddress]
+		if endpointRef == "" {
+			d.lc.Warnf("Registered device %s is missing required %s protocol information: %s.",
+				dev.Name, OnvifProtocol, EndpointRefAddress)
+			continue
+		}
+
+		deviceMap[endpointRef] = dev
+	}
+
+	return deviceMap
+}
+
+// discoverFilter iterates through the discovered devices, and returns any that are not duplicates
+// of devices in metadata or are from an alternate discovery method
+// will return an empty slice if no new devices are discovered
+func (d *Driver) discoverFilter(discoveredDevices []sdkModel.DiscoveredDevice) (filtered []sdkModel.DiscoveredDevice) {
+	// filter out duplicate discovered devices by endpoint reference
+	discoveredMap := make(map[string]sdkModel.DiscoveredDevice)
+	for _, device := range discoveredDevices {
+		endpointRef := device.Protocols[OnvifProtocol][EndpointRefAddress]
+		if _, found := discoveredMap[endpointRef]; !found {
+			discoveredMap[endpointRef] = device
+		}
+	}
+
+	existingDevices := d.makeDeviceMap() // create comparison map
+
+	// loop through discovered devices and see if they are already discovered
+	for endpointRef, device := range discoveredMap {
+		if existingDevice, found := existingDevices[endpointRef]; found {
+			if err := d.updateExistingDevice(existingDevice, device); err != nil {
+				d.lc.Errorf("error occurred while updating existing device %s: %s", existingDevice.Name, err.Error())
+			}
+			continue // skip registering existing device
+		}
+		// if device was not found, add it to the list of new devices to be registered with EdgeX
+		filtered = append(filtered, device)
+	}
+
+	return filtered
+}
+
+// updateExistingDevice compares a discovered device and a matching existing device, and updates the existing
+// device network address and port if necessary
+func (d *Driver) updateExistingDevice(device models.Device, discDev sdkModel.DiscoveredDevice) error {
+	shouldUpdate := false
+	if device.OperatingState == models.Down {
+		device.OperatingState = models.Up
+		shouldUpdate = true
+	}
+
+	device.Protocols[OnvifProtocol][LastSeen] = time.Now().Format(time.UnixDate)
+
+	existAddr := device.Protocols[OnvifProtocol][Address]
+	existPort := device.Protocols[OnvifProtocol][Port]
+	discAddr := discDev.Protocols[OnvifProtocol][Address]
+	discPort := discDev.Protocols[OnvifProtocol][Port]
+	if existAddr != discAddr ||
+		existPort != discPort {
+		d.lc.Info("Existing device has been discovered with a different network address.",
+			"oldInfo", fmt.Sprintf("%+v", existAddr+":"+existPort),
+			"discoveredInfo", fmt.Sprintf("%+v", discAddr+":"+discPort))
```

Comment:
there was a comment i left about turning this into an `Infof` call, and making the values be `%s` and not `%+v`.

Reply:
That's my bad I messed up part of a merge I think